Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,207 @@
+
+=============
+Ancillary Bus
+=============
+
+See <linux/ancillary_bus.h> for the models for ancillary_device and
+ancillary_driver.
+
+In some subsystems, the functionality of the core device (PCI/ACPI/other) may
+be too complex for a single device to be managed as a monolithic block or
+a part of the functionality might need to be exposed to a different subsystem.
+Splitting the functionality into smaller orthogonal devices would make it
+easier to manage data, power management and domain-specific interaction with
+the hardware. A key requirement for such a split is that there is no dependency
+on a physical bus, device, register accesses or regmap support. These
+individual devices split from the core cannot live on the platform bus as they
+are not physical devices that are controlled by DT/ACPI. The same argument
+applies for not using MFD in this scenario as MFD relies on individual function
+devices being physical devices that are DT enumerated.
+
+An example for this kind of requirement is the audio subsystem where a single
+IP handles multiple entities such as HDMI, Soundwire, local devices such as
+mics/speakers etc. The split for the core's functionality can be arbitrary or
+be defined by the DSP firmware topology and include hooks for test/debug. This
+allows for the audio core device to be minimal and focused on hardware-specific
+control and communication.
+
+The ancillary bus is intended to be minimal, generic and avoid domain-specific
+assumptions. Each ancillary_device represents a part of its parent
+functionality. The generic behavior can be extended and specialized as needed
+by encapsulating an ancillary_device within other domain-specific structures and
+the use of .ops callbacks. Devices on the ancillary bus do not share any
+structures and the use of a communication channel with the parent is
+domain-specific.
+
+When Should the Ancillary Bus Be Used
+=====================================
+
+When a kernel driver registers either an ancillary_device or ancillary_driver on
+the ancillary bus, we can use the nomenclature to refer to this kernel driver as
+a registering driver.  It is the entity that will allocate memory for either an
+ancillary_device or ancillary_driver and register the allocated object on the
+ancillary bus.
+
+One example could be a multi-port PCI network interface card that will allocate
+and register an ancillary_device for each physical function on the NIC.  To be
+matched with these ancillary_devices, a RDMA driver will allocate and register
+an ancillary_driver that will get matched with and probed for each of the
+ancillary_devices.
+
+The ancillary bus is to be used when multiple registering drivers, who share a
+common header file, need a mechanism to connect and provide access to a shared
+object allocated by the ancillary_device's registering driver.  The registering
+drivers for the ancillary_devices and ancillary_drivers can be from the same
+subsystem, or from multiple subsystems.
```

Comment:
This paragraph captures the use-case of multiple network drivers registering ancillary device for binding to a single rdma driver via a generic bus. And that registering driver and ancillary driver can be in different subsystem. Thats good but why did we choose ancillary bus for this is not clear.
I think we need to motivate that the PCI core device is managed by the network driver and it wants to export the rdma functionality to the rdma ancillary driver. This is why ancillary bus is a good fit.

This might need more shaping. But want to outline my thoughts. So something like --

"The registering driver and the ancillary_driver can be within the same subsystem, or from multiple subsystems.
One example is if multiple network drivers that manage core PCI devices wants to expose its rdma block functionality to a single rdma driver and provide a domain specific shared object.
The network drivers, i.e. registering drivers, allocates an ancillary device, encapsulate it within a domain specific shared object and register the ancillary devices for the PCI core devices. The ancillary bus will match the ancillary devices to a registered ancillary rdma driver. The ancillary device and its encapsulating shared object is available to rdma driver at probe time"

Reply:
Hi Dave - Ignore this comment for now. I realized that latest doc has some of this clipped. So I will revisit this section and comment after you update with Parav's feedback too. Sorry for the thrash!